### PR TITLE
Add the possibility to create custom LinearTasks in python

### PR DESCRIPTION
**Pull request by GiulioRomualdi**
If you are a `python` user and you want to create inverse kinematics (IK) or a task-based inverse
dynamics (TSID) task, you do need to write a `C++` code. Indeed it is possible to write a class that inherits from `TSIDLinearTask` or `IKLinearTask`  and passes it to `tsid` or `ik` solvers. The following snippet can be used as starting point to build a custom TSID task, a similar approach can be used to define a custom IK task

```py
import bipedal_locomotion_framework.bindings as blf


class CustomTSIDTask(blf.tsid.TSIDLinearTask):
    """ CustomTSIDTask represents a custom task for the tsid solver.

    The class must define the following functions:
    __init__(self): the constructor
    size(self) -> Get the size of the task. (I.e the number of rows of the vector b)
    is_valid(self) -> Determines the validity of the objects retrieved with getA() and getB()
    type(self) -> Get the type of the task. Namely equality or inequality

    Furthemore you have to populate the following attrubutes defined in TSIDLinearTask:
    self._description:str -> contains the description of the task
    self._A:numpy.ndarray[float64[m, n]]) -> matrix that describes the task A * x = b
    self._b: numpy.ndarray[float64[m, 1]]) -> vector that describes the task A * x = b

    Depending on the content of the task you may need to reimplement the following methods
    initialize(self, param_handler: blf.parameters_handler.IParameterHandler) -> usefull function to set parameters to the task
    set_variables_handler(self, variables_handler: blf.system.VariablesHandler) -> usefull function to set the variables
    update(self) -> update the content of the task
    """

    def __init__(self):

        blf.tsid.TSIDLinearTask.__init__(self)  # Without this, a TypeError is raised.

        # set the description of the task
        self._description = "Custom task"

    def size(self):
        # this is the number of rows of the vector b (in this example is equal to 1)
        return 1

    def is_valid(self):
        return True

    def type(self):
        # the types are defined in blf.tsid.TSIDLinearTask.Type
        return blf.tsid.TSIDLinearTask.Type.equality
```

Then you can use your custom task as follows
```py
import bipedal_locomotion_framework.bindings as blf

# Instantiate the task
custom_task = CustomTSIDTask()

# Initialize your task
custom_task.initialize(param_handler)

# Create the solver
solver = blf.tsid.QPTSID()
priority = 0
solver.add_task(custom_task, 'custom_task', priority) 
```

### CODE REVIEW

**Comment by GiulioRomualdi** [comment]
On conda macos there is this nice failure
```
../bindings/python/TSID/tests/test_TSID.py Fatal Python error: Illegal instruction

Current thread 0x000000011c3e5dc0 (most recent call first):
  File "/usr/local/miniconda/envs/test/lib/python3.9/site-packages/_pytest/python.py", line 183 in pytest_pyfunc_call
  File "/usr/local/miniconda/envs/test/lib/python3.9/site-packages/pluggy/_callers.py", line 39 in _multicall
  File "/usr/local/miniconda/envs/test/lib/python3.9/site-packages/pluggy/_manager.py", line 80 in _hookexec
  File "/usr/local/miniconda/envs/test/lib/python3.9/site-packages/pluggy/_hooks.py", line 265 in __call__
  File "/usr/local/miniconda/envs/test/lib/python3.9/site-packages/_pytest/python.py", line 1641 in runtest
  File "/usr/local/miniconda/envs/test/lib/python3.9/site-packages/_pytest/runner.py", line 162 in pytest_runtest_call
  File "/usr/local/miniconda/envs/test/lib/python3.9/site-packages/pluggy/_callers.py", line 39 in _multicall
  File "/usr/local/miniconda/envs/test/lib/python3.9/site-packages/pluggy/_manager.py", line 80 in _hookexec
  File "/usr/local/miniconda/envs/test/lib/python3.9/site-packages/pluggy/_hooks.py", line 265 in __call__
  File "/usr/local/miniconda/envs/test/lib/python3.9/site-packages/_pytest/runner.py", line 255 in <lambda>
  File "/usr/local/miniconda/envs/test/lib/python3.9/site-packages/_pytest/runner.py", line 311 in from_call
  File "/usr/local/miniconda/envs/test/lib/python3.9/site-packages/_pytest/runner.py", line 254 in call_runtest_hook
  File "/usr/local/miniconda/envs/test/lib/python3.9/site-packages/_pytest/runner.py", line 215 in call_and_report
  File "/usr/local/miniconda/envs/test/lib/python3.9/site-packages/_pytest/runner.py", line 126 in runtestprotocol
  File "/usr/local/miniconda/envs/test/lib/python3.9/site-packages/_pytest/runner.py", line 109 in pytest_runtest_protocol
  File "/usr/local/miniconda/envs/test/lib/python3.9/site-packages/pluggy/_callers.py", line 39 in _multicall
  File "/usr/local/miniconda/envs/test/lib/python3.9/site-packages/pluggy/_manager.py", line 80 in _hookexec
  File "/usr/local/miniconda/envs/test/lib/python3.9/site-packages/pluggy/_hooks.py", line 265 in __call__
  File "/usr/local/miniconda/envs/test/lib/python3.9/site-packages/_pytest/main.py", line 348 in pytest_runtestloop
  File "/usr/local/miniconda/envs/test/lib/python3.9/site-packages/pluggy/_callers.py", line 39 in _multicall
  File "/usr/local/miniconda/envs/test/lib/python3.9/site-packages/pluggy/_manager.py", line 80 in _hookexec
  File "/usr/local/miniconda/envs/test/lib/python3.9/site-packages/pluggy/_hooks.py", line 265 in __call__
  File "/usr/local/miniconda/envs/test/lib/python3.9/site-packages/_pytest/main.py", line 323 in _main
  File "/usr/local/miniconda/envs/test/lib/python3.9/site-packages/_pytest/main.py", line 269 in wrap_session
  File "/usr/local/miniconda/envs/test/lib/python3.9/site-packages/_pytest/main.py", line 316 in pytest_cmdline_main
  File "/usr/local/miniconda/envs/test/lib/python3.9/site-packages/pluggy/_callers.py", line 39 in _multicall
  File "/usr/local/miniconda/envs/test/lib/python3.9/site-packages/pluggy/_manager.py", line 80 in _hookexec
  File "/usr/local/miniconda/envs/test/lib/python3.9/site-packages/pluggy/_hooks.py", line 265 in __call__
  File "/usr/local/miniconda/envs/test/lib/python3.9/site-packages/_pytest/config/__init__.py", line 162 in main
  File "/usr/local/miniconda/envs/test/lib/python3.9/site-packages/_pytest/config/__init__.py", line 185 in console_main
  File "/usr/local/miniconda/envs/test/lib/python3.9/site-packages/pytest/__main__.py", line 5 in <module>
  File "/usr/local/miniconda/envs/test/lib/python3.9/runpy.py", line 87 in _run_code
  File "/usr/local/miniconda/envs/test/lib/python3.9/runpy.py", line 197 in _run_module_as_main
```

**Comment by GiulioRomualdi** [comment]
> Probably related to [conda/conda#9678](https://github.com/conda/conda/issues/9678) . Not sure what a quick workaround is.

Why is this happening now?

**Comment by DanielePucci** [comment]
CC @paolo-viceconte @diegoferigo @CarlottaSartore

**Comment by traversaro** [comment]
> > Probably related to [conda/conda#9678](https://github.com/conda/conda/issues/9678) . Not sure what a quick workaround is.
> 
> Why is this happening now?

It is possible that the other tests were not using `numpy` or they were not using some specific `numpy` function that is instead used by this test?

**Comment by GiulioRomualdi** [comment]
> It is possible that the other tests were not using numpy or they were not using some specific numpy function that is instead used by this test?

well `np.zeros()`, slicing and `np.ones()` were not used before 

```python
temp = np.zeros((10, variables_handler.get_number_of_variables()))
temp[:, variable.offset] = np.ones(10)
self._b = np.ones(10)
```

**Comment by traversaro** [comment]
> > It is possible that the other tests were not using numpy or they were not using some specific numpy function that is instead used by this test?
> 
> well `np.zeros()`, slicing and `np.ones()` were not used before
> 
> ```python
> temp = np.zeros((10, variables_handler.get_number_of_variables()))
> temp[:, variable.offset] = np.ones(10)
> self._b = np.ones(10)
> ```

I suspect this is what is triggering it. Note that I guess this is a problem that will not occur on real machines, just something that is occuring with virtual machine that do not support AVX instructions. For the time being you can just skip those tests on macOS.

File: docs/pages/python-additional-info.md [hunk 1]
```diff
@@ -60,6 +61,73 @@ vector = [1, 2, 3, 4, 5]
 ~~~~~~~~~~~~~
 
 ## Some utilities
+
+### Create a custom TSID or IK task directly in Python
+If you are a `python` user and you want create a inverse kinematics (IK) or a task based inverse
+dynamics (TSID) tasks, you do need to write a `C++` code. Indeed it is possible to write a class that inherits from `TSIDLinearTask` or `IKLinearTask`  and pass it to `tsid` or `ik` solvers. The following snippet can be used as starting point to build a custom TSID task, a similar approach can be used to define a custom IK task
```
Comment (paolo-viceconte) [inline docs/pages/python-additional-info.md:67]: Should this be "you don't need to write a `C++` code"?

**Review by paolo-viceconte — APPROVED** [review]
Looks good to me.

**Comment by traversaro** [comment]
In https://github.com/ami-iit/bipedal-locomotion-framework/pull/481 I tried to switch the blas implementation to netlib and blis, but the issue is still there, so I guess it is not a problem of blas implementation.

**Comment by GiulioRomualdi** [comment]
Bingo this is the problem:

I tested with https://github.com/GiulioRomualdi/blf-fastmac

```py
blf.tsid.TSIDLinearTask.__init__(self)
```

however, I don't know how to fix it since that line is required as explained in https://pybind11.readthedocs.io/en/stable/advanced/classes.html#overriding-virtual-functions-in-python

**Comment by GiulioRomualdi** [comment]
```
(robotologyenv) Mac-1638874300540:build runner$ python3 test.py
[2021-12-07 11:21:46.641] [thread: 106743] [blf] [info] [QPTSID::initialize] 'verbosity' not found. T
he following parameter will be used 'false'.
Illegal instruction: 4
```

https://stackoverflow.com/questions/14268887/what-is-the-illegal-instruction-4-error-and-why-does-mmacosx-version-min-10

**Comment by traversaro** [comment]
Are you sure that `python3` is the one provided in the conda environment? What is the output of `which python3`?

**Comment by GiulioRomualdi** [comment]
> Are you sure that `python3` is the one provided in the conda environment? What is the output of `which python3`?

```
/Users/runner/miniforge3/envs/robotologyenv/bin/python3
```

**Comment by traversaro** [comment]
In the recent past we had an "Illegal instruction" error for what in the end was just wrong memory access (see https://github.com/robotology/yarp-telemetry/issues/110). Did you tried to run python that runs this test on Linux under valgrind?

**Review by traversaro — APPROVED** [review]

**Comment by traversaro** [comment]
However, at this point for me we can disable the conda/macOS test and proceed.

**Comment by GiulioRomualdi** [comment]
That one was actually the problem, the test runs on `fastmac`

**Pull request merged**